### DISCUSSION TITLE
raspberry pi pico w using gpio input via interrupt while running a webserver

### DISCUSSION THREAD

**Discussion by raspberrypieman** [micropython/micropython#13665] [RP2040 / Pico] [answered]
I am trying to get input from a TCRT5000 infra-red detector on GPIO pin 15, via an interrupt routine - that works OK.  Then I added code to use the socket library to run small webserver and display the condition of the GPIO pin detected in the interrupt routine, via a global variable. 
The webserver works fine, until the first trigger of the interrupt routine - then it responds no more, however the interrupt routine, still works.
Is what I am attempting to do even possible on a Raspberry Pi Pico W ?

**Comment by sophiedegran** [comment]
Hi
Did you use asyncio mode for your project ?
It seem to be that the interrupt routine didn't releases to your web server after done..
Have a look to microdot it's small web server (all in async) which can do that well (look at examples dir)
Hope that help

**Reply by raspberrypieman** [reply]
Thank-you for your response. I did not use asyncio mode in my project - don't know how. However, I will look into microdot - do you have a link I can follow, please?

**Reply by sophiedegran** [reply]
-> github.com/miguelgrinberg/microdot

**Reply by bixb922** [reply, marked as answer]
And for asyncio, see https://github.com/peterhinch/micropython-async

Interrupts can be tricky. To check a sensor it's normally easier to start a second asyncio task with asyncio.create_task, with a loop to poll the sensor and asyncio.sleep_ms between 10 ms to 1 sec or more depending on what you want to achieve. This task will then run in parallel to the microdot server

**Reply by raspberrypieman** [reply]
Thanks, sophiedegran and bixb922. I will investigate and report back.

**Reply by raspberrypieman** [reply]
I managed to get a version working using asyncio for webserver and GPIO interrupt for TCRT5000 sensor - thanks to  sophiedegran and bixb922.
